src/main/java/com/limechain/storage/block/BlockState.java

```diff
@@ -0,0 +1,798 @@
+        return foundHeader.isPresent();
+    }
+
+    /**
+     * Get the block header for a given hash
+     *
+     * @param hash the hash of the block header in byte array representation
+     * @return the block header
+     */
+    public BlockHeader getHeader(byte[] hash) {
+        Block block = unfinalizedBlocks.get(hash);
+        if (block != null && block.getHeader() != null) {
+            return block.getHeader();
+        }
+
+        return loadHeaderFromDatabase(hash);
+    }
+
+    /**
+     * Get the block hash on our best chain with the given number
+     *
+     * @param blockNum the block number
+     * @return the block hash in byte array representation
+     */
+    public byte[] getHashByNumber(BigInteger blockNum) {
+        try {
+            // Try to get the hash from the block tree
+            return blockTree.getHashByNumber(blockNum.longValue());
+        } catch (RuntimeException e) {
+            // If the error is not LowerThanRootException, rethrow it
+            if (!(e instanceof LowerThanRootException)) {
+                throw new RuntimeException("Failed to get hash from blocktree: " + e.getMessage(), e);
+            }
+
+            // If error is LowerThanRootException, number has already been finalized, so check db
+            byte[] hash = (byte[]) db.find(helper.headerHashKey(blockNum)).orElse(null);
+
+            if (hash == null) {
+                throw new RuntimeException("Block " + blockNum + " not found");
+            }
+
+            return hash;
+        }
+    }
+
+    /**
+     * Get the block hash on our best chain with the given number
+     *
+     * @param blockNumber the block number
+     * @return List of block hashes in byte array representation
+     */
+    public List<byte[]> getHashesByNumber(BigInteger blockNumber) {
+        Block block;
+        try {
+            block = getBlockByNumber(blockNumber);
+        } catch (Exception e) {
+            throw new RuntimeException("Getting block by number: " + e.getMessage(), e);
+        }
+
+        List<byte[]> blockHashes = blockTree.getAllBlocksAtNumber(block.getHeader().getParentHash().getBytes());
+
+        byte[] hash = block.getHeader().getHash();
+        if (!blockHashes.contains(hash)) {
+            blockHashes.add(hash);
+        }
+
+        return blockHashes;
+    }
+
+    /**
+     * Get all the descendants for a given block hash (including itself), by first checking in memory
+     * and, if not found, reading from the block state database
+     *
+     * @param hash the hash of the block header in byte array representation
+     * @return List of block hashes in byte array representation
+     */
+    public List<byte[]> getAllDescendants(byte[] hash) {
+        List<byte[]> allDescendants;
+        try {
+            return blockTree.getAllDescendants(hash);
+        } catch (Exception e) {
+            if (!(e instanceof IllegalArgumentException)) {
+                throw e;
+            }
```

Comment:
I don't like the use of `instanceof` to check the type of exception. What's the problem with using multiple catch blocks to handle the different exceptions differently?